src/ssh/cockpitsshrelay.c

```diff
@@ -1320,61 +1320,46 @@ send_auth_reply (CockpitSshData *data,
   return ret;
 }
 
-static void
+static gboolean
 parse_host (const gchar *host,
             gchar **hostname,
             gchar **username,
             guint *port)
 {
-  gchar *user_arg = NULL;
-  gchar *host_arg = NULL;
-  gchar *tmp = NULL;
-  gchar *end = NULL;
-
-  guint64 tmp_num;
-
-  gsize host_offset = 0;
-  gsize host_length = strlen (host);
+  gchar *user_end = NULL;
+  g_autoptr(GError) error = NULL;
+  g_autoptr(GNetworkAddress) address = NULL;
 
-  tmp = strrchr (host, '@');
-  if (tmp)
+  user_end = strrchr (host, '@');
+  if (user_end)
     {
-      if (tmp[0] != host[0])
+      if (user_end != host)
         {
-          user_arg = g_strndup (host, tmp - host);
-          host_offset = strlen (user_arg) + 1;
-          host_length = host_length - host_offset;
+          *username = g_strndup (host, user_end - host);
+          host = user_end + 1;
         }
       else
         {
           g_message ("ignoring blank user in %s", host);
         }
     }
-
-  tmp = strrchr (host, ':');
-  if (tmp)
+  else
     {
-      tmp_num = g_ascii_strtoull (tmp + 1, &end, 10);
-      if (end[0] == '\0' && tmp_num > 0 && tmp_num < G_MAXUSHORT)
-        {
-          *port = (guint) tmp_num;
-          host_length = host_length - strlen (tmp);
-        }
-      else
-        {
-          g_message ("ignoring invalid port in %s", host);
-        }
+      *username = g_strdup (g_get_user_name ());
     }
 
-  if (!user_arg || user_arg[0] == '\0')
-    user_arg = (gchar *) g_get_user_name ();
-
-  host_arg = g_strndup (host + host_offset, host_length);
-  *hostname = g_strdup (host_arg);
-  *username = g_strdup (user_arg);
-
-  g_free (host_arg);
-  g_free (user_arg);
+  address = G_NETWORK_ADDRESS (g_network_address_parse (host, 0, &error));
```

Comment:
Can you use a static (C-style) cast here?  It would make me happier for two silly reasons :)

Reply:
Done. OOI, why? The dynamic glib ones are safer and fail much more usefully, and this code isn't at all performance critical.

Reply:
Basically, the fancy cast has no value here (since it's a constructor and we know exactly what type we're getting out of it) and because I always forget if we're allowed to put NULL through a GLib cast or not :rofl: